src/Microsoft.AspNetCore.Routing/Matchers/DataSourceDependentMatcher.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    internal class DataSourceDependentMatcher : Matcher
+    {
+        private readonly Func<MatcherBuilder> _matcherBuilderFactory;
+        private readonly DataSourceDependentCache<Matcher> _cache;
+
+        public DataSourceDependentMatcher(
+            EndpointDataSource dataSource,
+            Func<MatcherBuilder> matcherBuilderFactory)
+        {
+            _matcherBuilderFactory = matcherBuilderFactory;
+
+            _cache = new DataSourceDependentCache<Matcher>(dataSource, CreateMatcher);
+            _cache.EnsureInitialized();
```

Comment:
Is it worth benchmarking how long it takes to initialize the matcher? re: startup time and time to first request